src/libre/ast.c

```diff
@@ -912,3 +913,128 @@ ast_node_type_name(enum ast_expr_type t)
 	case AST_EXPR_TOMBSTONE: return "TOMBSTONE";
 	}
 }
+
+/*
+ * TODO: we could use this for AST->NFA conversion and decide to call
+ * re_strings() instead (also for a set of alts of strings)
+ *
+ * Return values:
+ *  -1: Error
+ *   0: Not a literal, *s and *n are not defined. *anchor_start/end are not defined.
+ *   1: Literal, *s may or may not be NULL (i.e. for an empty string), *n >= 0
+ *      and *anchor_start/_end are set.
+ */
+int
+ast_expr_is_literal(const struct ast_expr *e,
+	int *anchor_start, int *anchor_end,
+	char **s, size_t *n)
+{
+	size_t count;
+	const struct ast_expr **nodes;
+	int is_end_nl;
+	size_t i;
+
+	is_end_nl = 0;
+
+	*anchor_start = 0;
+	*anchor_end   = 0;
+
+	/* correctly match the empty string */
+	if (e == NULL || e->type == AST_EXPR_EMPTY) {
+		count = 0;
+		nodes = NULL;
+		goto done;
+	}
+
+	/*
+	 * TODO: We could also handle AST_EXPR_CODEPOINT here
+	 * (and in concatenations below), and perhaps render out
+	 * the string to utf8.
+	 */
+
+	/* a single character has no concat node (or anchors) */
+	if (e->type == AST_EXPR_LITERAL) {
+		count = 1;
+		nodes = &e;
+		goto done;
+	}
+
+	if (e->type != AST_EXPR_CONCAT && e->type != AST_EXPR_ANCHOR) {
+		return 0;
+	}
+
+	/*
+	 * At this point we're considering either a single anchor,
```

Comment:
Since this (quite reasonably, imo) only accepts single anchors, unlike PCRE, is it worth adding a test for e.g. `^^^abc$$$` being rejected?

Reply:
Good call, I merged #399 and added a test in caa41834